internal/controllers/machinedeployment/machinedeployment_rollout_sequence_test.go

```diff
@@ -0,0 +1,876 @@
+			indent := "  "
+			if strings.HasPrefix(line, "[") {
+				indent = ""
+			}
+			sb.WriteString(indent + line + "\n")
+		}
+	} else {
+		sb.WriteString(s + "\n")
+	}
+	l.testCaseStringBuilder.WriteString(sb.String())
+}
+
+func (l *fileLogger) EndTestCase() (string, string, error) {
+	if err := os.WriteFile(fmt.Sprintf("%s.test.log", l.fileName), []byte(l.testCaseStringBuilder.String()), 0600); err != nil {
+		return "", "", err
+	}
+	if err := os.WriteFile(fmt.Sprintf("%s.test.log.golden", l.fileName), []byte(l.testCaseStringBuilder.String()), 0600); err != nil {
+		return "", "", err
+	}
+
+	currentBytes, _ := os.ReadFile(fmt.Sprintf("%s.test.log", l.fileName))
+	current := string(currentBytes)
+
+	goldenBytes, _ := os.ReadFile(fmt.Sprintf("%s.test.log.golden", l.fileName))
+	golden := string(goldenBytes)
+
+	return current, golden, nil
+}
+
+func sortMachineSetMachines(machines []*clusterv1.Machine) {
+	sort.Slice(machines, func(i, j int) bool {
+		iIndex, _ := strconv.Atoi(strings.TrimPrefix(machines[i].Name, "m"))
+		jiIndex, _ := strconv.Atoi(strings.TrimPrefix(machines[j].Name, "m"))
+		return iIndex < jiIndex
+	})
+}
+
+func minAvailableBreachToleration() func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+	return func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+		log.Logf("[Toleration] tolerate minAvailable breach")
+		return true
+	}
+}
+
+func maxSurgeToleration() func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+	return func(log *fileLogger, _ int, _ *rolloutScope, _, _ int32) bool {
+		log.Logf("[Toleration] tolerate maxSurge breach")
+		return true
+	}
+}
+
+// default task order ensure the controllers are run in a consistent and predictable way: md, ms1, ms2 and so on.
+func defaultTaskOrder(current *rolloutScope) []int {
+	taskOrder := []int{}
+	for t := range len(current.machineSets) + 1 + 1 { // +1 is for the MachineSet that might be created when reconciling md, +1 is for the md itself
+		taskOrder = append(taskOrder, t)
+	}
+	return taskOrder
+}
+
+func randomTaskOrder(current *rolloutScope, rng *rand.Rand) []int {
+	u := &UniqueRand{
+		rng:       rng,
+		generated: map[int]bool{},
+		max:       len(current.machineSets) + 1 + 1, // +1 is for the MachineSet that might be created when reconciling md, +1 is for the md itself
+	}
+	taskOrder := []int{}
+	for {
+		n := u.Int()
+		if u.rng.Int()%10 < 3 { // skip a step in the 30% of cases
+			continue
+		}
+		taskOrder = append(taskOrder, n)
+		if r := u.rng.Int() % 10; r < 3 { // repeat a step in the 30% of cases
+			delete(u.generated, n)
+		}
+		if len(u.generated) >= u.max {
+			break
+		}
+	}
+	return taskOrder
+}
+
+func createMD(failureDomain string, replicas int32, maxSurge, maxUnavailable int32) *clusterv1.MachineDeployment {
+	return &clusterv1.MachineDeployment{
+		ObjectMeta: metav1.ObjectMeta{Name: "md"},
+		Spec: clusterv1.MachineDeploymentSpec{
+			// Note: using failureDomain as a template field to determine upToDete
```

Comment:
```suggestion
			// Note: using failureDomain as a template field to determine upToDate
```
(same below two more times)